Fixed: Playing video is too slow to start.

MohitMaliFtechiz opened the pull request:
Fixes #3909 

* Previously the `data` has been read for the video within the application for setting the content range, which slowdowns the process of video loading so we have refactored our code to not use that, and should directly pass the stream for video to webView so that webView handled the stream itself.
* Rendering the HTML data, and video data on the IO thread.
* Now we are passing the stream if any uncompressed item over 1MB comes to render, and loading the data via libkiwix for the rest.

kelson42 commented:
It should not only "reduce" the video loading. The loading of a video should be immediate as there is no reason that it takes any time.

You don't need the data to have its size. Just use the `item.getSize()`, see https://libzim.readthedocs.io/en/latest/api/classzim_1_1Item.html#classzim_1_1Item

From the code it seems as well as video are treated exactly like other kind of content. Do you confirm?

MohitMaliFtechiz commented:
> It should not only "reduce" the video loading. The loading of a video should be immediate as there is no reason that it takes any time.

The video is immediately available as it comes from an offline source. it just takes 50-100MS to read and prepare the inputStream from the video file. 

> You don't need the data to have its size. Just use the item.getSize(), see https://libzim.readthedocs.io/en/latest/api/classzim_1_1Item.html#classzim_1_1Item

@kelson42 Here `data` is content that we are loading for rendering in the webView. For this type of video URL, we already have the content length which comes from this `data` variable, and we were again getting the length via the `item.getSize()` method which takes more time to return the response to webView since there is one additional unnecessary request to libzim.

```kotlin
 fun load(url: String, requestHeaders: Map<String, String>): WebResourceResponse {
    val data = zimFileReader?.load(url)
    return WebResourceResponse(
      zimFileReader?.getMimeTypeFromUrl(url),
      Charsets.UTF_8.name(),
      data
    )
      .apply {
        val headers = mutableMapOf("Accept-Ranges" to "bytes")
        if ("Range" in requestHeaders.keys) {
          setStatusCodeAndReasonPhrase(HttpURLConnection.HTTP_PARTIAL, "Partial Content")
          val fullSize = when {
            // check if the previously loaded data has the size then return it.
            // It will prevent the again data loading for those videos that are already loaded.
            // See #3909
            data?.available() != null && data.available().toLong() != 0L ->
              data.available().toLong()

            else -> {
              // if the loaded data does not have the size, especially for YT videos.
              // Then get the content size from libzim and set it to the headers.
              zimFileReader?.getItem(url)?.size ?: 0L
            }
          }
          val lastByte = fullSize - 1
          val byteRanges = requestHeaders.getValue("Range").substringAfter("=").split("-")
          headers["Content-Range"] = "bytes ${byteRanges[0]}-$lastByte/$fullSize"
          if (byteRanges.size == 1) {
            headers["Connection"] = "close"
          }
        } else {
          setStatusCodeAndReasonPhrase(HttpURLConnection.HTTP_OK, "OK")
        }
        responseHeaders = headers
      }
  }
```

![Screenshot from 2024-07-09 18-19-49](https://github.com/kiwix/kiwix-android/assets/34593983/009b2ad5-26ad-4845-8b31-5f37ca0ae1fb)


The main reason for the delay in playing videos is due to the loading of closed caption (CC) lists, which can take up to 6-7 seconds for videos like 'The transformative power of classical music,' which has the largest CC list. This loading process occurs in the background thread to ensure smooth scrolling for users. The video does not start playing immediately in the WebView because there are pending requests already in the WebView's processing queue. WebView handles URLs sequentially, so when we attempt to play the video, it joins the queue. Once the pending CC list requests are complete, the play video request is processed, and the video begins playing. This is the limitation in the webView that we can not load the requests on the different threads because the `shouldInterceptRequest` method gives the URL from the queue. However, the webView requests are handled on the `worker thread` but it is a signal thread so it manages the queue for processing the requests.

On the other hand, in IOS we also implemented this type of approach(https://github.com/kiwix/kiwix-apple/pull/774) where the loading of content is moved to the background thread. The video loading progress bar is showing until the content is not fully loaded.

![Screenshot from 2024-07-09 18-11-51](https://github.com/kiwix/kiwix-android/assets/34593983/f8418410-8355-434e-a61e-91d3216e0d23)

On the kiwix browser(extension of Chrome), there is no limitation of this type which is why we can immediately process the next request(playing video on click) on the other background thread(or they are probably using a new `Promise` for every request that makes the request completion independently from the other request).

kelson42 commented:
> @kelson42 Here `data` is content that we are loading for rendering in the webView. For this type of video URL, we already have the content length which comes from this `data` variable, and we were again getting the length via the `item.getSize()` method which takes more time to return the response to webView since there is one additional unnecessary request to libzim.

You should not read yourself the data for the video... think what will happen if the video makes 2GB! All of this was working fine, since when do we read the full data of a video ourself?

MohitMaliFtechiz commented:
@kelson42 It was introduced here https://github.com/kiwix/kiwix-android/commit/8498a6b1ef15c12a890f2aa5dcc535793a25e9a2 and removed in https://github.com/kiwix/kiwix-android/commit/42015d5cb0089e9f15b989843492a0dbe1b17a3b, and again we have introduced it in https://github.com/kiwix/kiwix-android/commit/3388142142f4d77b06db6991c8889ddc008ca377 for fixing the video seeking problem.

kelson42 commented:
Please secure that any uncompressed file over 1MB is passed a stream a viewer should deal with it... If we still face regressions them they should be fixed properly.

MohitMaliFtechiz commented:
I have made the changes in the code to ensure that the stream should pass for the uncompressed file. I also moved the video loading to the background thread.

kelson42 commented:
> I have made the changes in the code to ensure that the stream should pass for the uncompressed file and Kiwix does not do anything with that. I also moved the video loading to the background thread.

Sorry for my ignorance but where is this decided exactly? this is very not clear. Please add also a limit based on the siye: continue to load the data via libkiwix if the siye < 1MB

MohitMaliFtechiz commented:
> Sorry for my ignorance but where is this decided exactly? this is very not clear. Please add also a limit based on the siz: continue to load the data via libkiwix if the size < 1MB

We already passing the stream for the videos to webView, and for the rest of the content I have added the limit based on the size, if any uncompressed item over 1MB comes then it will return the stream, and for the rest content, the data will load via the libkiwix.
